### PR TITLE
[IBS-69] Test & migrate groups (student)

### PR DESCRIPTION
**Pull request by hezhongq**
## Description
- Jira: https://ibs-utm.atlassian.net/browse/IBS-69
- Refactored Group endpoints for Students.

## Changes Made
- Refactored following endpoints`
  - `module/group/student/accept.js`
  - `module/group/student/check.js`
  - `module/group/student/create.js`
  - `module/group/student/disinvite.js`
  - `module/group/student/invite.js`
  - `module/group/student/leave.js`
  - `module/group/student/reject.js`
- Enabled the disinvite endpoint for a student in `backend/route/student.js`for refactoring purposes

## Test Guidelines
- In `/backend/` run the following and ensure that they all passes:
 ```bash 
npm test group_student_accept
```
 ```bash 
npm test group_student_check
```
 ```bash 
npm test group_student_create
```
 ```bash 
npm test group_student_disinvite
```
 ```bash 
npm test group_student_invite
```
 ```bash 
npm test group_student_leave
```
 ```bash 
npm test group_student_reject
```

### For PR author:
- [ ] Reviewed code changes.
- [ ] Tested changes locally.
- [ ] Added appropriate comments.
- [ ] Ensured consistent code style and formatting.
- [ ] Included unit/integration tests.
- [ ] Handled edge cases appropriately.
- [ ] Celebrated accomplishments 🎉

### For PR reviewer:
- [ ] Reviewed code changes.
- [ ] Tested changes locally.
- [ ] Verified clear and helpful comments.
- [ ] Ensured code style aligns with standards.
- [ ] Examined test coverage.
- [ ] Provided feedback on potential edge cases.

### CODE REVIEW

File: backend/lib/version_control/MockVCS.js [hunk 1]
```diff
@@ -29,18 +30,19 @@ class MockVCS extends AbstractVCS {
...
         // 1. calls get_vcs_user_id(username) to get user_id
-        let user_id = await this.get_vcs_user_id(username);
+        let user_id = await MockVCS.get_user_id(username);
 
         if (user_id === -1) {
             return { success: false, code: 'gitlab_invalid_username' };
         }
 
-        return { success: true, gitlab_url: vcs_url };
+        return { success: true, url: vcs_url };
```
Comment (DomiVesalius) [inline backend/lib/version_control/MockVCS.js:39]: I remember helping you debug an issue where the url was not showing up in the response body.

This return value needs to be exactly the same as the one in `GitlabVCS` so you cannot change it from `gitlab_url` to `url`.  Change this back to `gitlab_url` and modify any places that make use of this return value accordingly.

File: backend/migrations/20231008081024-create-course.js [hunk 2]
```diff
@@ -60,6 +60,6 @@ module.exports = {
   async down(queryInterface, Sequelize) {
     await queryInterface.dropAllTables();
     // Remove the foreign key constraint
-    await queryInterface.removeConstraint('courses', 'custom_fkey_course_id');
+    // await queryInterface.removeConstraint('courses', 'custom_fkey_course_id');
```
Comment (DomiVesalius) [inline backend/migrations/20231008081024-create-course.js:63]: Remove commented code

File: backend/module/group/student/accept.js [hunk 3]
```diff
@@ -1,44 +1,66 @@
 const express = require("express");
 const router = express.Router();
-const client = require("../../../setup/db");
-const helpers = require("../../../utilities/helpers");
+const { QueryTypes } = require('sequelize');
```
Comment (DomiVesalius) [inline backend/module/group/student/accept.js:3]: Remove unused import

File: backend/module/group/student/check.js [hunk 4]
```diff
@@ -1,56 +1,56 @@
...
-		return;
-	}
+router.get("/", async (req, res) => {
+	try {
 
-	let sql_select_group = "SELECT * FROM course_" + res.locals["course_id"] + ".group_user WHERE username = ($1) AND task = ($2)";
-	let sql_select_members = "SELECT username, status FROM course_" + res.locals["course_id"] + ".group_user WHERE group_id = ($1)";
-	let sql_select_gitlab_url = "SELECT * FROM course_" + res.locals["course_id"] + ".group WHERE group_id = ($1)";
+		const { username } = res.locals;
+		console.log({ username });
```
Comment (DomiVesalius) [inline backend/module/group/student/check.js:9]: Remove log

File: backend/module/group/student/check.js [hunk 5]
```diff
@@ -1,56 +1,56 @@
...
 const router = express.Router();
-const client = require("../../../setup/db");
-const helpers = require("../../../utilities/helpers");
+const { GroupUser, Group } = require("../../../models");
 
-router.get("/", (req, res) => {
-	if (res.locals["task"] === "") {
-		res.status(400).json({ message: "The task is missing or invalid." });
-		return;
-	}
```
Comment (DomiVesalius) [inline backend/module/group/student/check.js:10]: I think you still need to check for this case.

File: backend/module/group/student/check.js [hunk 6]
```diff
@@ -1,56 +1,56 @@
...
 			res.status(200).json({ message: "You are not in a group." });
+			return;
 		}
-	});
-})
 
-module.exports = router;
\ No newline at end of file
+		const group_id = groupUser.dataValues.group_id;
+		const status = groupUser.dataValues.status;
```
Comment (DomiVesalius) [inline backend/module/group/student/check.js:24]: remove unused variable

File: backend/module/group/student/check.js [hunk 7]
```diff
@@ -1,56 +1,56 @@
...
-	if (res.locals["interview_group"] !== "" && res.locals["interview_group"] !== null) {
-		var task = res.locals["interview_group"];
-	} else {
-		var task = res.locals["task"];
-	}
+		// Find the GroupUser record for the given username and task
+		const groupUser = await GroupUser.findOne({
+			where: { username },
+			attributes: ['task_id', 'username', 'group_id', 'status']
+		});
```
Comment (DomiVesalius) [inline backend/module/group/student/check.js:15]: I don't think this block of code is the correct translation of the SQL statement it is trying to mimic:
```sql
SELECT * FROM course_" + res.locals["course_id"] + ".group_user WHERE username = ($1) AND task = ($2)
```

You forgot to specify the task in the `where` clause. Currently, if the user was in multiple groups across several courses with several tasks, then this might end up choosing the incorrect group that you are checking for.

File: backend/module/group/student/create.js [hunk 8]
```diff
@@ -2,58 +2,58 @@ const express = require("express");
...
-			} else {
-				res.status(404).json({ message: "Unknown error." });
-				console.log(err);
-			}
-		} else if (pgRes.rowCount === 1) {
-			let group_id = pgRes.rows[0]["group_id"];
-			client.query(sql_add_user, [res.locals["task"], res.locals["username"], group_id], (err, pgRes) => {
-				if (err) {
-					if (err.code === "23505") {
-						res.status(409).json({ message: "You can join at most one group for each task." });
```
Comment (DomiVesalius) [inline backend/module/group/student/create.js:33]: The refactored version should include this logic as well

File: backend/module/group/student/create.js [hunk 9]
```diff
@@ -2,58 +2,58 @@ const express = require("express");
...
-				}
-			});
+		if (!taskInstance) {
+			return res.status(400).json({ message: "The task is not found in the database." });
 		}
-	});
-})
+
+		// Create a new group using the found task ID
+		const newGroup = await Group.create({ task_id: taskInstance.id });
```
Comment (DomiVesalius) [inline backend/module/group/student/create.js:28]: Refer to previous comment. You must check if the user is already in a group before creating a new one.

File: backend/module/group/student/disinvite.js [hunk 10]
```diff
@@ -1,47 +1,54 @@
+// routes/group/disinvite.js
 const express = require("express");
 const router = express.Router();
-const client = require("../../../setup/db");
+const { GroupUser, Group } = require('../../../models');
```
Comment (DomiVesalius) [inline backend/module/group/student/disinvite.js:4]: Remove unused `Group` model from import

File: backend/module/group/student/disinvite.js [hunk 11]
```diff
@@ -1,47 +1,54 @@
...
-	client.query(sql_select_user, [res.locals["task"] , res.locals["username"]], (err, pgRes) => {
-		if (err) {
-			res.status(404).json({ message: "Unknown error." });
-			console.log(err);
-			return;
-		} else if (pgRes.rowCount !== 1) {
-			res.status(403).json({ message: "You don't have access to cancel the invitation." });
-			return;
+		const { username } = res.locals;
+		const { task } = res.locals.task;
```
Comment (DomiVesalius) [inline backend/module/group/student/disinvite.js:22]: Unused variable

File: backend/module/group/student/disinvite.js [hunk 12]
```diff
@@ -1,47 +1,54 @@
...
+
+		// Check if the user has access to cancel the invitation
+		const groupUser = await GroupUser.findOne({
+			where: { username: username },
+			attributes: ['task_id', 'username', 'group_id', 'status']
+		});
```
Comment (DomiVesalius) [inline backend/module/group/student/disinvite.js:28]: Same issue pointed out by previous comment. I don't think this is a direct translation of
```sql
SELECT * FROM course_" + res.locals["course_id"] + ".group_user WHERE task = ($1) AND username = ($2) AND status = 'confirmed'
```

File: backend/module/group/student/invite.js [hunk 13]
```diff
@@ -1,73 +1,55 @@
...
+		if (!taskInstance) {
+			return res.status(400).json({ message: "The task is not found in the database." });
+		}
 
-	let sql_select_users = "SELECT * FROM course_" + res.locals["course_id"] + ".group_user WHERE task = ($1) AND username = ($2)";
-	let sql_select_max_member = "SELECT max_member FROM course_" + res.locals["course_id"] + ".task WHERE task = ($1)";
-	let sql_invite = "INSERT INTO course_" + res.locals["course_id"] + ".group_user (task, username, group_id, status) VALUES (($1), ($2), ($3), 'pending')";
+		const usersInGroup = await GroupUser.findAll({
+			where: { task_id: taskInstance.id, status: 'confirmed' }
+		});
```
Comment (DomiVesalius) [inline backend/module/group/student/invite.js:21]: This isn't a direct translation of
```sql
SELECT * FROM course_" + res.locals["course_id"] + ".group_user WHERE task = ($1) AND username = ($2)
```

You need to include username in your where clause otherwise all this statement does is get every confirmed group_user row for a specific task.

File: backend/module/group/student/invite.js [hunk 14]
```diff
@@ -1,73 +1,55 @@
...
+			return res.status(400).json({ message: "No more user can be invited as the maximum has been reached." });
 		}
-		if (!has_access) {
-			res.status(400).json({ message: "You don't have access to invite." });
-			return;
+
+		const { username: invitedUsername } = req.body;
+		if (!invitedUsername || helpers.name_validate(invitedUsername)) {
+			return res.status(400).json({ message: "The username is missing or has an invalid format." });
 		}
```
Comment (DomiVesalius) [inline backend/module/group/student/invite.js:37]: Should also check if a user with the given username exists in the database to cover this case from the original version: https://github.com/Kianoosh76/IBS/blob/d3317ecf1d68513b00df81e966ee0c42b356d373/backend/module/group/student/invite.js#L58

File: backend/module/group/student/reject.js [hunk 15]
```diff
@@ -1,30 +1,50 @@
...
+			return res.status(400).json({ message: "The task is missing or invalid." });
+		}
+
+		// Find the task ID based on the provided criteria
+		const taskDetails = await Task.findOne({
+			where: { course_id, task }
+		});
+
+		if (!taskDetails) {
+			return res.status(400).json({ message: "Task not found." });
```
Comment (DomiVesalius) [inline backend/module/group/student/reject.js:29]: Original message should be preserved:

https://github.com/Kianoosh76/IBS/blob/d3317ecf1d68513b00df81e966ee0c42b356d373/backend/module/group/student/reject.js#L12

**Review by DomiVesalius — CHANGES_REQUESTED** [review]
## Issues
There are a few issues in the way previous SQL queries have been translated to ORM calls that need to be fixed.

Also noticed that this should include refactor for `module/group/student/leave.js` however it remains unchanged. I see that you have pushed a test and seeder file for that endpoint so I am assuming you simply forgot to push those changes.

## Tests
- Tests for accept, check, create, disinvite, invite, and reject pass
- Test for leave fails due to missing refactored version

File: backend/module/group/student/invite.js (hunk 13 above)
Comment (hezhongq) [reply backend/module/group/student/invite.js:21]: I have taken a different approach of first finding all users in the group, then create invitation if the following check passes:
1. current user has access to invite other user(user who initiates the invitation is already in the group)
2. if usersInGroup.length has not yet reach max number of available user per group
This approach helps with getting the headcount of students per group

File: backend/module/group/student/invite.js (hunk 14 above)
Comment (hezhongq) [reply backend/module/group/student/invite.js:37]: When we reach this line it would have already passed the username check, where if the task with given username does not exist then it could just return with the "You don't have access to invite." message instead. We may not necessarily need to reveal much info regarding our data in the database

**Comment by mohtashim-ali** [comment]
I have reviewed the PR and it seems to be completed, just needs to be merged. All tests pass with latest changes from previous PRs merged.

**Review by Kianoosh76 — APPROVED** [review]
